File: 3.math-stats/subsets/subsets.rb
def subsets(nums)
  res = []
  dfs(nums, [], res)
  return res
end

def dfs(nums, path, res)
  res.push(path)
  for i in 0..(nums.length - 1)
    dfs(nums[i + 1..-1], path + [nums[i]], res)
  end
  return res
end

nums = [1, 2, 3]
p(subsets(nums))

File: 3.math-stats/subsets/subsets.js
const subsets = function (nums) {
  res = [];
  dfs(nums, [], res);
  return res;
};

const dfs = function (nums, path, res) {
  res.push(path);
  for (let i = 0; i < nums.length; ++i) {
    const newNums = [...nums];
    newNums.splice(0, i + 1);
    dfs(newNums, [...path, ...[nums[i]]], res);
  }
  return res;
};
const nums = [1, 2, 3];
console.log(subsets(nums));
